plugins/experimental/ja3_fingerprint/ja3_fingerprint.cc

```diff
@@ -0,0 +1,489 @@
+/** @ja3_fingerprint.cc
+  Plugin JA3 Fingerprint calculates JA3 signatures for incoming SSL traffic.
+  @section license License
+  Licensed to the Apache Software Foundation (ASF) under one
+  or more contributor license agreements.  See the NOTICE file
+  distributed with this work for additional information
+  regarding copyright ownership.  The ASF licenses this file
+  to you under the Apache License, Version 2.0 (the
+  "License"); you may not use this file except in compliance
+  with the License.  You may obtain a copy of the License at
+      http://www.apache.org/licenses/LICENSE-2.0
+  Unless required by applicable law or agreed to in writing, software
+  distributed under the License is distributed on an "AS IS" BASIS,
+  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  See the License for the specific language governing permissions and
+  limitations under the License.
+ */
+
+#include <inttypes.h>
+#include <cstdlib>
+#include <cstdio>
+#include <cstring>
+#include <cerrno>
+#include <stdlib.h>
+#include <getopt.h>
+#include <netinet/in.h>
+#include <arpa/inet.h>
+
+#include <algorithm>
+#include <fstream>
+#include <sstream>
+#include <string>
+#include <unordered_set>
+#include <unordered_map>
+#include <memory>
+#include <regex>
+
+#include "ts/ts.h"
+#include "ts/remap.h"
+
+#include <openssl/ssl.h>
+#include <openssl/md5.h>
+#include <openssl/opensslv.h>
+
+// Get 16bit big endian order and update pointer
+#define n2s(c, s) ((s = (((unsigned int)(c[0])) << 8) | (((unsigned int)(c[1])))), c += 2)
+
+const char *PLUGIN_NAME = "ja3_fingerprint";
+static TSTextLogObject pluginlog;
+static int ja3_idx    = -1;
+static int enable_raw = 0;
+static int enable_log = 0;
+
+// GREASE table as in ja3
+static const std::unordered_set<uint16_t> GREASE_table = {0x0a0a, 0x1a1a, 0x2a2a, 0x3a3a, 0x4a4a, 0x5a5a, 0x6a6a, 0x7a7a,
+                                                          0x8a8a, 0x9a9a, 0xaaaa, 0xbaba, 0xcaca, 0xdada, 0xeaea, 0xfafa};
+
+struct ja3_data {
+  std::string ja3_string;
+  char md5String[33];
+  char ip_addr[INET6_ADDRSTRLEN];
+};
+
+struct ja3_remap_info {
+  int raw        = false;
+  int log        = false;
+  TSCont handler = nullptr;
+
+  ~ja3_remap_info()
+  {
+    if (handler) {
+      TSContDestroy(handler);
+      handler = nullptr;
+    }
+  }
+};
+
+static int
+custom_get_ja3_prefixed(int unit, const unsigned char *&data, int len, std::string &result)
+{
+  int cnt, tmp;
+  bool first = true;
+  // Extract each entry and append to result string
+  for (cnt = 0; cnt < len; cnt += unit) {
+    if (unit == 1) {
```

Comment:
Can `unit` only be the values 1 or 2? If so it would be better as an enum, e.g. "enum Unit { Byte, Short}". Although looking at the code, the loop constructs might well be different enough to justify two methods, one for each of the two sizes.